File: tests/commissioning_spi_v1/test_spi_v1_protocol.py
"""[COMMISSIONING_SPI_V1] Protocol commissioning tests for BPM <-> UCE SPI V1."""

import unittest


MAGIC = 0x53
VERSION = 0x01
FRAME_DATA = 0x01
FRAME_ACK = 0x02
ACK = 0x01
RETRY = 0x02
DROP = 0x03


def crc8(data):
    crc = 0
    for value in data:
        crc ^= value
        for _ in range(8):
            crc = ((crc << 1) ^ 0x07) & 0xFF if crc & 0x80 else (crc << 1) & 0xFF
    return crc


def cobs_encode(data):
    out = bytearray([0])
    code_index = 0
    code = 1
    for value in data:
        if value == 0:
            out[code_index] = code
            code_index = len(out)
            out.append(0)
            code = 1
        else:
            out.append(value)
            code += 1
            if code == 0xFF:
                out[code_index] = code
                code_index = len(out)
                out.append(0)
                code = 1
    out[code_index] = code
    return bytes(out)


def cobs_decode(data):
    out = bytearray()
    index = 0
    while index < len(data):
        code = data[index]
        if code == 0:
            raise ValueError("zero inside COBS payload")
        index += 1
        end = index + code - 1
        if end > len(data):
            raise ValueError("COBS block overruns payload")
        out.extend(data[index:end])
        index = end
        if code != 0xFF and index < len(data):
            out.append(0)
    return bytes(out)


def tlv_packet(t, value):
    return bytes([t, len(value)]) + bytes(value)


def data_frame(message_id, *tlvs):
    frame = bytearray([MAGIC, VERSION, FRAME_DATA, len(tlvs)])
    for offset, tlv in enumerate(tlvs):
        mid = (message_id + offset) & 0xFF
        item = bytes([mid]) + tlv
        frame.extend(item)
        frame.append(crc8(item))
    assert len(frame) <= 256
    return bytes(frame)


def ack_frame(*items):
    frame = bytearray([MAGIC, VERSION, FRAME_ACK, len(items)])
    for message_id, status in items:
        item = bytes([message_id, status])
        frame.extend(item)
        frame.append(crc8(item))
    assert len(frame) <= 256
    return bytes(frame)


def parse_data(frame):
    assert frame[:3] == bytes([MAGIC, VERSION, FRAME_DATA])
    count = frame[3]
    cursor = 4
    parsed = []
    while cursor < len(frame):
        message_id = frame[cursor]
        cursor += 1
        length = frame[cursor + 1]
        tlv_len = length + 2
        tlv = frame[cursor : cursor + tlv_len]
        msg_crc = frame[cursor + tlv_len]
        assert crc8(bytes([message_id]) + tlv) == msg_crc
        parsed.append((message_id, bytes(tlv)))
        cursor += tlv_len + 1
    assert len(parsed) == count
    return parsed


class SpiV1ProtocolTests(unittest.TestCase):
    def test_frame_data_multiple_messages_round_trip(self):
        frame = data_frame(7, tlv_packet(0x12, [1]), tlv_packet(0x22, []), tlv_packet(0x20, [3, 4, 5]))
        physical = cobs_encode(frame) + b"\x00"
        decoded = cobs_decode(physical[:-1])
        self.assertEqual(parse_data(decoded), [(7, tlv_packet(0x12, [1])), (8, tlv_packet(0x22, [])), (9, tlv_packet(0x20, [3, 4, 5]))])

    def test_ack_statuses_validate_crc(self):
        frame = ack_frame((1, ACK), (2, RETRY), (3, DROP))
        self.assertEqual(frame[2], FRAME_ACK)
        for cursor in range(4, len(frame), 3):
            self.assertEqual(crc8(frame[cursor : cursor + 2]), frame[cursor + 2])

    def test_priority_cycle_is_6_3_1(self):
        cycle = ["A"] * 6 + ["M"] * 3 + ["B"]
        self.assertEqual(cycle, ["A", "A", "A", "A", "A", "A", "M", "M", "M", "B"])

    def test_no_fragmentation_or_skip_when_next_message_does_not_fit(self):
        frame = data_frame(1, tlv_packet(0x10, [0x55] * 240))
        next_msg = tlv_packet(0x11, [0x66] * 8)
        self.assertGreater(len(frame) + 1 + len(next_msg) + 1, 256)

    def test_timeout_policy_is_500_ms_without_progress(self):
        last_progress_ms = 1000
        now_ms = 1501
        self.assertGreater(now_ms - last_progress_ms, 500)

    def test_window_lifecycle_ack_retry_drop(self):
        states = {1: "SENT_WAIT_ACK", 2: "SENT_WAIT_ACK", 3: "SENT_WAIT_ACK"}
        for message_id, status in [(1, ACK), (2, RETRY), (3, DROP)]:
            if status == ACK:
                states[message_id] = "DONE"
            elif status == RETRY:
                states[message_id] = "RETRY_PENDING"
            elif status == DROP:
                states[message_id] = "DISCARDED"
        self.assertEqual(states, {1: "DONE", 2: "RETRY_PENDING", 3: "DISCARDED"})

    def test_pause_resume_model_blocks_clock_while_irq_high(self):
        irq_levels = ["LOW", "HIGH", "HIGH", "LOW"]
        clocks = ["CLK" if level == "LOW" else "PAUSE" for level in irq_levels]
        self.assertEqual(clocks, ["CLK", "PAUSE", "PAUSE", "CLK"])

    def test_header_is_read_before_payload_without_releasing_cs(self):
        response = tlv_packet(0x12, [0x01]) + bytes([crc8(tlv_packet(0x12, [0x01]))])
        cs_levels = ["LOW", "LOW", "LOW", "LOW"]
        header = response[:2]
        payload = response[2:]
        self.assertEqual(header, bytes([0x12, 0x01]))
        self.assertEqual(payload, bytes([0x01, response[-1]]))
        self.assertTrue(all(level == "LOW" for level in cs_levels))


if __name__ == "__main__":
    unittest.main()

File: tests/commissioning_spi_v1/test_spi_v1_protocol.mjs
// [COMMISSIONING_SPI_V1] Executable protocol checks for environments with Node.js.
import assert from "node:assert/strict";

const MAGIC = 0x53;
const VERSION = 0x01;
const FRAME_DATA = 0x01;
const FRAME_ACK = 0x02;
const ACK = 0x01;
const RETRY = 0x02;
const DROP = 0x03;

function crc8(data) {
  let crc = 0;
  for (const value of data) {
    crc ^= value;
    for (let bit = 0; bit < 8; bit += 1) {
      crc = (crc & 0x80) ? ((crc << 1) ^ 0x07) & 0xff : (crc << 1) & 0xff;
    }
  }
  return crc;
}

function cobsEncode(data) {
  const out = [0];
  let codeIndex = 0;
  let code = 1;
  for (const value of data) {
    if (value === 0) {
      out[codeIndex] = code;
      codeIndex = out.length;
      out.push(0);
      code = 1;
    } else {
      out.push(value);
      code += 1;
      if (code === 0xff) {
        out[codeIndex] = code;
        codeIndex = out.length;
        out.push(0);
        code = 1;
      }
    }
  }
  out[codeIndex] = code;
  return Uint8Array.from(out);
}

function cobsDecode(data) {
  const out = [];
  let index = 0;
  while (index < data.length) {
    const code = data[index++];
    assert.notEqual(code, 0);
    const end = index + code - 1;
    assert.ok(end <= data.length);
    while (index < end) out.push(data[index++]);
    if (code !== 0xff && index < data.length) out.push(0);
  }
  return Uint8Array.from(out);
}

function tlvPacket(type, value) {
  return Uint8Array.from([type, value.length, ...value]);
}

function dataFrame(messageId, ...tlvs) {
  const frame = [MAGIC, VERSION, FRAME_DATA, tlvs.length];
  tlvs.forEach((tlv, offset) => {
    const item = Uint8Array.from([(messageId + offset) & 0xff, ...tlv]);
    frame.push(...item, crc8(item));
  });
  assert.ok(frame.length <= 256);
  return Uint8Array.from(frame);
}

function ackFrame(...items) {
  const frame = [MAGIC, VERSION, FRAME_ACK, items.length];
  for (const [id, status] of items) {
    const item = Uint8Array.from([id, status]);
    frame.push(...item, crc8(item));
  }
  assert.ok(frame.length <= 256);
  return Uint8Array.from(frame);
}

function parseData(frame) {
  assert.deepEqual([...frame.slice(0, 3)], [MAGIC, VERSION, FRAME_DATA]);
  const count = frame[3];
  const parsed = [];
  let cursor = 4;
  while (cursor < frame.length) {
    const id = frame[cursor++];
    const len = frame[cursor + 1];
    const tlvLen = len + 2;
    const tlv = frame.slice(cursor, cursor + tlvLen);
    const msgCrc = frame[cursor + tlvLen];
    assert.equal(crc8(Uint8Array.from([id, ...tlv])), msgCrc);
    parsed.push([id, [...tlv]]);
    cursor += tlvLen + 1;
  }
  assert.equal(parsed.length, count);
  return parsed;
}

const frame = dataFrame(7, tlvPacket(0x12, [1]), tlvPacket(0x22, []), tlvPacket(0x20, [3, 4, 5]));
const physical = Uint8Array.from([...cobsEncode(frame), 0]);
assert.deepEqual(parseData(cobsDecode(physical.slice(0, -1))), [
  [7, [...tlvPacket(0x12, [1])]],
  [8, [...tlvPacket(0x22, [])]],
  [9, [...tlvPacket(0x20, [3, 4, 5])]],
]);
assert.equal(frame.length, 4 + (1 + tlvPacket(0x12, [1]).length + 1) + (1 + tlvPacket(0x22, []).length + 1) + (1 + tlvPacket(0x20, [3, 4, 5]).length + 1));
assert.equal(1 + tlvPacket(0x12, [1]).length + 1, 1 + 2 + 1 + 1);

const acks = ackFrame([1, ACK], [2, RETRY], [3, DROP]);
for (let cursor = 4; cursor < acks.length; cursor += 3) {
  assert.equal(crc8(acks.slice(cursor, cursor + 2)), acks[cursor + 2]);
}

assert.deepEqual(["A", "A", "A", "A", "A", "A", "M", "M", "M", "B"], [...Array(6).fill("A"), ...Array(3).fill("M"), "B"]);
const largeFrame = dataFrame(1, tlvPacket(0x10, Array(240).fill(0x55)));
assert.ok(largeFrame.length + 1 + tlvPacket(0x11, Array(8).fill(0x66)).length + 1 > 256);
assert.equal(parseData(largeFrame).length, 1);
assert.ok(1501 - 1000 > 500);

const states = new Map([[1, "SENT_WAIT_ACK"], [2, "SENT_WAIT_ACK"], [3, "SENT_WAIT_ACK"]]);
for (const [id, status] of [[1, ACK], [2, RETRY], [3, DROP]]) {
  states.set(id, status === ACK ? "DONE" : status === RETRY ? "RETRY_PENDING" : "DISCARDED");
}
assert.deepEqual(Object.fromEntries(states), { 1: "DONE", 2: "RETRY_PENDING", 3: "DISCARDED" });
assert.deepEqual(["LOW", "HIGH", "HIGH", "LOW"].map((level) => level === "LOW" ? "CLK" : "PAUSE"), ["CLK", "PAUSE", "PAUSE", "CLK"]);
const response = Uint8Array.from([...tlvPacket(0x12, [0x01]), crc8(tlvPacket(0x12, [0x01]))]);
assert.deepEqual([...response.slice(0, 2)], [0x12, 0x01]);
assert.deepEqual([...response.slice(2)], [0x01, response[response.length - 1]]);
assert.deepEqual(["LOW", "LOW", "LOW", "LOW"], Array(4).fill("LOW"));

console.log("[COMMISSIONING_SPI_V1] protocol checks passed");
